Sync `v2-3-stable` with `v2-3-test` to release `2.3.3`

ephraimbuddy opened the pull request:
Time for `2.3.3rc1`!

GApfel committed:
Allow for LOGGING_LEVEL=DEBUG (#23360)

(cherry picked from commit c83d0bfef6664b4c8579ebb91c2af1fac669c2ca)

bbovenzi committed:
Maintain grid view selection on filtering upstream (#23779)

* Maintain grid selection on filter upstream

The grid view selection was being cleared when clicking "Filter Upstream". The selection should persist.

Also, added a left margin to the "Reset root" button

* fix linting

(cherry picked from commit 9db2271fad751b51127ac0c2dc0477dc95b271d5)

mfjackson committed:
Alphabetizes two tables (#23923)

The rest of the page has consistently alphabetized tables. This commit fixes three `extras` that were not alphabetized.

(cherry picked from commit ef3f5301f3f04b6718b2aeb722deffd9c075fcde)

pierrejeambrun committed:
Grid fix details button truncated and small UI tweaks (#23934)

* Show details button and wrap on LegendRow.

* Update following brent review

* Fix display on small width

* Rotate icon for a 'ReadLess' effect

(cherry picked from commit 5ab58d057abb6b1f28eb4e3fb5cec7dc9850f0b0)

sekikn committed:
Fix inverted section levels in best-practices.rst (#23968)

This PR fixes inverted levels in the sections added to the "Best Practices" document in #21879.

(cherry picked from commit 8e7b76de9a726a8d085fe5b875331b66bf3cd045)

ephraimbuddy committed:
Add TaskInstance State 'REMOVED' to finished states and success states (#23797)

Now that we support dynamic task mapping, we should have the 'REMOVED'
state of task instances as a finished state because
for dynamic tasks with a removed task instance, the dagrun would be stuck in
running state if 'REMOVED' state is not in finished states.

(cherry picked from commit 73446f28e9eb1e4c6f2f32c700147b61ab3da600)

eladkal committed:
Replace `use_task_execution_date` with `use_task_logical_date` (#23983)

* Replace `use_task_execution_date` with `use_task_logical_date`
We have some operators/sensors that use `*_execution_date` as the class parameters. This PR deprecate the usage of these parameters and replace it with `logical_date`.
There is no change in functionality, under the hood the functionality already uses `logical_date` this is just about the parameters name as exposed to the users.

(cherry picked from commit 614b2329c1603ef1e2199044e2cc9e4b7332c2e0)

eladkal committed:
Cleanup `BranchDayOfWeekOperator` example dag (#24007)

* Cleanup BranchDayOfWeekOperator example dag
There is no need for `dag=dag` when using context manager.

(cherry picked from commit 56fd04016f1a8561f1c02e7f756bab8805c05876)

hubert-pietron committed:
Fix closing connection dbapi.get_pandas_df (#23452)

(cherry picked from commit ab1f637e463011a34d950c306583400b7a2fceb3)

bernardocouto committed:
Removing magic numbers from exceptions (#23997)

* Removing magic numbers from exceptions

* Running pre-commit

(cherry picked from commit 375673aa3cebf6889f403ac1577cffe0c33c1ae7)

bbovenzi committed:
Reduce grid view API calls (#24083)

* Reduce API calls from /grid

- Separate /grid_data from /grid
- Remove need for formatData
- Increase default query stale time to prevent extra fetches
- Fix useTask query keys

* consolidate grid data functions

* fix www tests

test grid_data instead of /grid

(cherry picked from commit 035553c86988f403b43ef5825715b45f055d62dd)

bernardocouto committed:
Removing magic status code numbers from api_connecxion (#24050)

(cherry picked from commit f2e6452efdf74828de9f73a8bf4f42f6dd10eb58)

jhtimmins committed:
Rename Permissions to Permission Pairs. (#24065)

(cherry picked from commit be21e08e1bb6202626c12b2375f24167cf22838a)

Commit:
Update production-deployment.rst (#24121)

The sql_alchemy_conn option is in the database section, not the core section.  Simple typo fix.

(cherry picked from commit f294a266de671514c5eaad183fe34028a97ccf1e)

potiuk committed:
Update 'rich' to latest version across the board. (#24186)

That Also includes regenerating the breeze output images.

(cherry picked from commit 13c62b52cbe0bdcf3319a403ce061268661245b3)

potiuk committed:
Add verification step to image release process (#24177)

(cherry picked from commit 5d5598eca6ae4513e2decd851d32db992b6a02ec)

potiuk committed:
Remove warnings when starting breeze (#24183)

Breeze when started produced three warnings that were harmless,
but we should fix them to remove "false positives".

(cherry picked from commit ac8a790ab9dabfa747964b65aa9b93451b735ede)

uranusjr committed:
Parse error for task added to multiple groups (#23071)

This raises an exception if a task already belonging to a task group
(including added to a DAG, since such task is automatically added to the
DAG's root task group).

Also, according to the issue response, manually calling TaskGroup.add()
is not considered a supported way to add a task to group. So a
meta-marker is added to the function docstring to prevent it from
showing up in documentation and users from trying to use it.

(cherry picked from commit 337863fa35bba8463d62e5cf0859f2bb73cf053a)

Bowrna committed:
Fix xfail test in test_scheduler.py (#23731)

(cherry picked from commit dbe80c89b2a99d6ab737f2c4146bf8f918034f0f)

bbovenzi committed:
Fix useTasks crash on error (#24152)

* Prevent UI from crashing on Get API error

* add test

* don't show API errors in test logs

* use setMinutes inline

(cherry picked from commit fd4e344880b505cfe53a97a6373d329515bbc7a3)

uranusjr committed:
Apply per-run log templates to log handlers (#24153)

(cherry picked from commit c23826915dcdca4f22b52b74633336cb2f4a1eca)

potiuk committed:
Fix choosing backend versions in breeze's command line (#24228)

Choosing version of backend were broken when command line switches
were used. The _VERSION variables were "hard-coded" to defaults
rather than taken from command line. This is a remnant of initial
implementation and converting the parameters to "cacheable" ones.

While looking at the versions we also found that PARAM_NAME_FLAG
is not used any more so we took the opportunity to remove it.

(cherry picked from commit 4482eaefb02535317bc0d2378eba6bfcef3c45ff)

potiuk committed:
Add command to regenerate breeze command output images (#24216)

(cherry picked from commit 2d687e4b7b39fc5c133440de7f66f58327d33db7)

bbovenzi committed:
Check that edge nodes actually exist (#24166)

(cherry picked from commit 3e51d8029ba34d3a76b3afe53e257f1fb5fb9da1)

malthe committed:
Mention context variables and logging (#24304)

* Mention context variables and logging

* Fix static checks

(cherry picked from commit 926d9bd1911adf21665910a514d5f214b4a093b6)

ephraimbuddy committed:
Fix permission issue for dag that has dot in name (#23510)

How we determine if a DAG is a subdag in airflow.security.permissions.resource_name_for_dag is not right.
If a dag_id contains a dot, the permission is not recorded correctly.

The current solution makes a query every time we check for permission for dags that has a dot in the name. Not that I like it but I think it's better than other options I considered such as changing how we name dags for subdag. That's not
good in UX. Another option I considered was making a query when parsing, that's not good and it's avoided
by passing root_dag to resource_name_for_dag

Co-authored-by: Ash Berlin-Taylor <ash_github@firemirror.com>
Co-authored-by: Tzu-ping Chung <uranusjr@gmail.com>
(cherry picked from commit cc35fcaf89eeff3d89e18088c2e68f01f8baad56)

uranusjr committed:
Check bag DAG schedule_interval match tiemtable (#23113)

This guards against the DAG's timetable or schedule_interval from being
changed after it's created. Validation is done by creating a timetable
and check its summary matches schedule_interval. The logic is not
bullet-proof, especially if a custom timetable does not provide a useful
summary. But this is the best we can do.

(cherry picked from commit a1a9a8f9a3adc63e783cf3fd699066f35e488d4f)

bbovenzi committed:
Check for run_id for grid group summaries (#24327)

(cherry picked from commit 29e10c5f026ad84c1ca4fb00c8f174f6eb3d640f)

bbovenzi committed:
Add typescript (#24337)

(cherry picked from commit 74ac9f788c31512b1fcd9254282905f34cc40666)

potiuk committed:
Add scripts that provide good links to example dags (#24348)

The documentation generated used "main" in the URL of the
example DAGs.

The generation of the links have been fixed in the #24307, but this
PR adds a tool that has been used to fix existing links in generated
documentation resulting in https://github.com/apache/airflow-site/pull/610

Fixes: #24331
(cherry picked from commit 85c247ae10da5ee93f26352d369f794ff4f2e47c)

jedcunningham committed:
Optimize calendar view for cron scheduled DAGs (#24262)

(cherry picked from commit 23fb6635ea31bc4d78b2bcdc07767dd8595a06af)

future-taga committed:
fix 2.3.2 release date. (#24370)

(cherry picked from commit 4daf51a2c388b41201a0a8095e0a97c27d6704c8)

ephraimbuddy committed:
Refactor `DagRun.verify_integrity` (#24114)

This refactoring became necessary as there's a necessity to add additional code
to the already exisiting code to handle mapped task immutability during run. The additional
code would make this method difficult to read. Refactoring the code will aid understanding and
help in debugging.

(cherry picked from commit 12638d2310d962986b43af8f1584a405e280badf)

ashb committed:
Add tests for the grid_data endpoint (#24375)

The one fix/change here was to include the JSON content response here so that
`resp.json` works in the test.

(cherry picked from commit 2b2d97068fa45881672dab6f2134becae246f3f3)

kinow committed:
Update Dockerfile (#24397)

(cherry picked from commit 5623a5f1bc83ddc026fd10b4ee14bd9856c2c98a)

potiuk committed:
Better diagnostics for ARM for MySQL and MSSQL (#24185)

Until we have debian suppor tof MySQL and MSSQL ARM, runnign
those on ARM platform is not supported. However error about it
was not clear (pulling docker image failed).

This PR adds platform checking also in breeze and fails fast
without even attempting to enter breeze shell when you are on
ARM and wants to run MsSQL or MySQL breeze shell.

Also some errors with running different backend versions via
breeze have been removed.

(cherry picked from commit 00d2a3c9319af6e145b8e03f5ac33a95fa8acd66)

potiuk committed:
Refresh list of committers (#24398)

(cherry picked from commit 4d2536584ba1e8fea79be93c009041d1d9deb933)

potiuk committed:
Fix flaky order of returned dag runs (#24405)

There was no ordering on a query returning dag_runs when it comes
to grid view. This caused flaky tests but also it would have
caused problems with random reordering of reported dagruns in the
UI (it seems).

This change adds stable ordering on returned Dag Runs:

* by dag_run_id (ascending) asc

No need to filter by map_index as there will be always max one
returned TI from each dag run

(cherry picked from commit 2edab57d4e8ccbd5b8f66c3951615c169fb0543e)

potiuk committed:
The timeouts for Helm tests are far too small for "full tests" (#24408)

When "Full tests needed" are run, then the Helm tests take far
more time because they are running mor combinations of executor
and Python version. Such tests will timeout now.

This PR increases the timeout

(cherry picked from commit 7729d1b60775197c321760e9a46f49e8cd3beac1)

potiuk committed:
Revert "Remove custom signal handling in Triggerer (#23274)" (#24390)

This reverts commit 6bdbed6c43df3c5473b168a75c50e0139cc13e88.

(cherry picked from commit 224285b6fcd433cf6257519f695d8d49feceebf2)

potiuk committed:
Fix links to sources for examples (#24386)

The links to example sources in exampleinclude have been broken in a
number of providers and they were additionally broken by AIP-47.

This PR fixes it.

Fixes: #23632
Fixes: https://github.com/apache/airflow-site/issues/536
(cherry picked from commit 08b675cf6642171cb1c5ddfb09607b541db70b29)

bbovenzi committed:
Improve grid rendering performance with a custom tooltip (#24417)

* fully custom tooltip

* use a customized chakra tooltip

* update notice

(cherry picked from commit e452949610cff67c0e0a9918a8fefa7e8cc4b8c8)

bbovenzi committed:
Upgrade to react 18 and chakra 2 (#24430)

* Upgrade to react 18 and chakra 2

* upgrade emotion cache

* fix framer-motion and (some) testing issues

(cherry picked from commit 6a1b323e0dec762e4d73a96b07427f3d7aadf4ba)

dojoscan committed:
Update errors.rst (#24412)

Fix wording in some sentences

(cherry picked from commit 5d5976c08c867b8dbae8301f46e0c422d4dde1ed)

fritz-astronomer committed:
`TI.log_url` fix for `map_index` (#24335)

(cherry picked from commit 48a6155bb1478245c1dd8b6401e4cce00e129422)

potiuk committed:
Add CI-friendly progress output for tests (#24236)

This is the first step to run breeze tests in parallel in CI.
This flag adds "limited progress" output when running tests
which means that the runnig tests will just print few lines with
percent progress and color status indication from last few
progress lines of Pytest output, but when it completes, the whole output is
printed in a CI group - colored depending on status.

The final version (wnen we implement parallel test execution) should
also defer writing the output to until all tests are completed, but
this should be a follow-up PR.

(cherry picked from commit 41fefa146ac64379447db503b7dba82d5121f06a)

jedcunningham committed:
Docs: Fix default 2.2.5 log_id_template (#24455)

I accidentally got the wrong default for 2.2.5 when documenting how
to fix elasticsearch remote logging after upgrading to 2.3.0+.

(cherry picked from commit c8fa9e96e29f9f8b4ff9c7db416097fb70a87c2d)

dstandish committed:
Don't crash scheduler if exec config has old k8s objects (#24117)

From time to time k8s library objects change their attrs.  If executor config is stored with old version, and unpickled with new version, we can get attribute errors that can crash the scheduler (see https://github.com/apache/airflow/issues/23727).

Here we update handling so that we fail the task but don't crash the scheduler.

(cherry picked from commit 0c41f437674f135fe7232a368bf9c198b0ecd2f0)

potiuk committed:
First attempt to have CI-controlled process of releasing PROD image (#24433)

(cherry picked from commit b79790d8e5de18782a3179a8a55457eafa529a88)

ashb committed:
Speed up grid_data endpoint by 10x (#24284)

* Speed up grid_data endpoint by 10x

These changes make the endpoint go from almost 20s down to 1.5s and the
changes are two fold:

1. Keep datetimes as objects for as long as possible

   Previously we were converting start/end dates for a task group to a
   string, and then in the parent parsing it back to a datetime to find
   the min and max of all the child nodes.

   The fix for that was to leave it as a datetime (or a
   pendulum.DateTime technically) and use the existing
   `AirflowJsonEncoder` class to "correctly" encode these objects on
   output.

2. Reduce the number of DB queries from 1 per task to 1.

   The removed `get_task_summaries` function was called for each task,
   and was making a query to the database to find info for the given
   DagRuns.

   The helper function now makes just a single DB query for all
   tasks/runs and constructs a dict to efficiently look up the ti by
   run_id.

* Add support for mapped tasks in the grid data

* Don't fail when not all tasks have a finish date.

Note that this possibly has incorrect behaviour, in that the end_date of
a TaskGroup is set to the max of all the children's end dates, even if
some are still running. (This is the existing behaviour and is not
changed or altered by this change - limiting it to just performance
fixes)

(cherry picked from commit 451a6f4d9ff8b744075e2f25099046c77f28179e)

bbovenzi committed:
Remove framer-motion from custom tooltip (#24449)

* remove framer-motion from custom tooltip

* remove framer-motion type declaration file

(cherry picked from commit 98c6fa32ba952f2611ceb3335da106a81b2a3e73)

jedcunningham committed:
Use `get_hostname` instead of `socket.getfqdn` (#24260)

We allow users to configure a different function to determine the
hostname, so we should use that consistently when we need the hostname.

(cherry picked from commit e29543ec00c0a3eae7a789cb49350499c3b584c2)

dstandish committed:
Return empty dict if Pod JSON encoding fails (#24478)

When UI unpickles executor_configs with outdated k8s objects it can run into the same issue as the scheduler does (see https://github.com/apache/airflow/issues/23727).

Our JSON encoder therefore needs to suppress encoding errors arising from pod serialization, and fallback to a default value.

josh-fell committed:
Small cleanup of ``get_current_context()`` chapter (#24482)

(cherry picked from commit 7848f0461c2ff254acceb7cd7cf5d7d900116499)

potiuk committed:
Improve production image release workflow (#24481)

Few improvements:

* direct link to workflow in the docs
* "green" success screenshots (separate screenshot for rc)
* job and step names contain both Airflow version and python version
* running subsequent build with same version will cancel past
  in-progress build (in case you quickly make constraint fis for
  example and re-run)

(cherry picked from commit 8ad18bf02924d1677502231e678951bc5b8cb420)

bbovenzi committed:
Remove internet explorer support (#24495)

* remove internet explorer support

* remove from webpack

(cherry picked from commit 05cc5ff73788dc9f1974df772f683020523092aa)

bbovenzi committed:
Upgrade to webpack 5 (#24485)

* update webpack-cli, eslint, stylelint, babel

* revert stylelint changes

* update more plugins

* update to webpack 5

* remove all resolutions

(cherry picked from commit 127f8f4de02422ade8f2c84f84d3262d6efde185)

dstandish committed:
Add indexes for CASCADE deletes for task_instance (#24488)

When we add foreign keys with ON DELETE CASCADE, and we delete rows in the foreign table, the database needs to join back to the referencing table.  If there's no suitable index, then it can be slow to perform the deletes.

(cherry picked from commit 677c42227c08f705142f298ab88915f133cd94e5)

bbovenzi committed:
Restore tooltip animation (#24503)

* update webpack-cli, eslint, stylelint, babel

* revert stylelint changes

* update more plugins

* update to webpack 5

* restore animation with framer-motion v6

(cherry picked from commit 9dd16be0c5459c5d438503ac1905716be04bad9a)

bbovenzi committed:
Fix toast messages (#24505)

* update webpack-cli, eslint, stylelint, babel

* revert stylelint changes

* update more plugins

* update to webpack 5

* use portalProps to properly style toast msgs

(cherry picked from commit c74946275e3488f3e553e156e8294cddae6b5db7)

jedcunningham committed:
Fix deprecated log_id_template value (#24506)

We had the wrong old default in our deprecated value upgrading logic, so
the 2.2.5 default wasn't actually being upgraded.

(cherry picked from commit a955ae4e7eef70a5621979152f9be2acb2211801)

jedcunningham committed:
Seed log_template table (#24511)

Seed the log_template table with the default values pre 2.3.0 so log
retrieval still works post upgrade. This only worked previously if you
have the default in your config, now it works even if you don't.

(cherry picked from commit c36d94df09930abfb58856e1b8b6e059c7a44525)

holadepo committed:
Update ui.rst (#24514)

Fix minor typos

(cherry picked from commit 31f113e4de019399ea6439909e5f6b1ad187fa3f)

bbovenzi committed:
Do not calculate grid root instances (#24528)

* do not calculate grid root instances

* fix tests

(cherry picked from commit 8b054efba736b985c8fdb7a278f01d5f4934f857)

potiuk committed:
Restore capability of reproducing CI failures using new Breeze (#24402)

The old breeze-legacy used to have a possibility of very easy
reproduction of CI failures by executing the right breeze command
that contained the commit hash of the PR being tested. This has
been broken for some time after we migrated to the new breeze,
but finally it was the time when it was needed again.

This PR brings back the capability by:

* addding --image-tag parameters for tests, shell and start-airflow
  commands
* if --image-tag is specified, then rather than building the
  image, it is pulled using the specified hash
* if --image-tag is specified, the local sources are not mounted
  to breeze when started, but the sources already embedded in the
  image are used ("skipped" set for --mount-sources).
* new "removed" command value is added to --mount-sources, it
  causes breeze command to remove the sources from the image (it
  is used when installing airflow during the tests for specified
  version (it's automatically used when --use-airflow-version
  is used).

(cherry picked from commit 7dc794a985c5e19126068ee1a83d09a21e894a3e)

ephraimbuddy committed:
Fix mapped task immutability after clear (#23667)

We should be able to detect if the structure of mapped task has changed
and verify the integrity.

This PR ensures this
Co-authored-by: Tzu-ping Chung <uranusjr@gmail.com>

(cherry picked from commit b692517ce3aafb276e9d23570e9734c30a5f3d1f)

josh-fell committed:
Move LOAD_DEFAULT_CONNECTIONS env var to database config section in CI (#24536)

(cherry picked from commit fe105f8887d516c4dc6ecff55dded2215280716c)

potiuk committed:
Get rid of TimedJSONWebSignatureSerializer (#24519)

The TimedJSONWebSignatureSerializer has been deprecated from the
itsdangerous library and they recommended to use dedicated
libraries for it.

https://github.com/pallets/itsdangerous/issues/129

Since we are going to move to FAB 4+ with #22397 where newer version of
itsdangerous is used, we need to switch to another library.

We are already using PyJWT so the choice is obvious.

Additionally to switching, the following improvements were done:

* the use of JWT claims has been fixed to follow JWT standard.
  We were using "iat" header wrongly. The specification of JWT only
  expects the header to be there and be valid UTC timestamp, but the
  claim does not impact maturity of the signature - the signature
  is valid if iat is in the future.
  Instead "nbf" - "not before" claim should be used to verify if the
  request is not coming from the future. We now require all claims
  to be present in the request.

* rather than using salt/signing_context we switched to standard
  JWT "audience" claim (same end result)

* we have now much better diagnostics on the server side of the
  reason why request is forbidden - explicit error messages
  are printed in server logs and details of the exception. This
  is secure, we do not spill the information about the reason
  to the client, it's only available in server logs, so there is
  no risk attacker could use it.

* the JWTSigner is "use-agnostic". We should be able to use the
  same class for any other signatures (Internal API from AIP-44)
  with just different audience

* Short, 5 seconds default clock skew is allowed, to account for
  systems that have "almost" synchronized time

* more tests addded with proper time freezing testing both
  expiry and immaturity of the request

This change is not a breaking one because the JWT authentication
details are not "public API" - but in case someone reverse engineered
our claims and implemented their own log file retrieval, we
should add a change in our changelog - therefore newsfragment
is added.

(cherry picked from commit 1f8e4c955394b31462956501d9a6741b98892f3a)

BasPH committed:
Add imports to deferring code samples (#24544)

(cherry picked from commit cabbf616bbc3139b1ce793c9daa20f13c82eb287)

potiuk committed:
Add verification steps when releasing the images. (#24520)

After the images are pushed in CI we are running the verification
of the AMD image now.

This cannot be really done during building and pushing the image,
because we are using multi-platform images using remote builders
so the image is not even available locally, so we need to actually
pull the images after they are built in order to verify them.

This PR adds those features:
* ability to pull images for verification with --pull-flag
* ability to verify slim images (regular tests are skipped and
  we only expect the preinstalled providers to be available
* the steps to verify the images (both regular and slim) are
  added to the workflow

(cherry picked from commit 2936759256273a66cf6343e451d81013a95ec47e)

potiuk committed:
Update flask-appbuilder authlib/oauth dependency (#24516)

The dependency we have for flask-appbuilder oauth authentication
(for github/google authentication) should follow the limits
that flask-appbuilder current version has. We added authlib there
but apparently FAB currently limits authlib to <= 1.0 - we should
follow fab rather than have our own dependency here.

This has been pointed out in
https://github.com/dpgaspar/Flask-AppBuilder/issues/1861

(cherry picked from commit 5674491dc8e8ed1685cdb4c04922cb72ad8ba9b4)

potiuk committed:
Update description of installing providers separately from core (#24454)

The description of how to install providers separately from the
core was wrong. It suggested installing main constraints, but
in fact we are not able to give any guarantees when installing
providers this way so no constraints can guarantee consistent
set of dependencies and the user is on their own in this case.

This PR corrects the installation docs.

(cherry picked from commit 28d3236202b34719020f9f7eb702af4d2f95a864)

potiuk committed:
Clarify that users should not use Maria DB (#24556)

(cherry picked from commit 05c542dfa8eee9b4cdca4e9370f459ce807354b2)

Commit:
Switch Markdown engine to markdown-it-py (#19702)

(cherry picked from commit 88363b543f6f963247c332e9d7830bc782ed6e2d)

potiuk committed:
Added instructions on what to do if your command images are regenerated (#24581)

In case there are conflicting changes to breeze command in several
PRs, you might get conflicting images printed. in such case you
should run `breeze regenerate-command-images`

(cherry picked from commit 864cbc9cd843db94ca8bab2187b50de714fdb070)

potiuk committed:
Switch to building images in parallell (#24580)

In the new Breeze, switching to using parallelism is a ... breeze.

This PR adds the capability of building the images in parallel in Breeze
locally - for breeze command, but also uses this capability to build the
images in parallel in our CI. Our builds are always executed on
powerful, big machines with lots of CPU and docker run in memory
filesystem with 32GB RAM, so it should be possible to run all builds in
parallel on a single machine rather then spin off parallel machines to
run the builds using the matrix strategy of Github Actions.

Generally speaking - this will either speed up or get 4x cost saving for
the build steps for all the "full test needed" PRs as well as all the
main builds.

There are a number of savings and improvements we can achieve this way:

1) less overhead for starting and runnning the machines
2) seems that with the new buildkit, the parallel builds are not
   suffering from some sequential locks (as it used to be, so
   we are basically do the same job using 25% resources for building
   the images.
3) we will stop having random "one image failed to build" cases - they
   will all either fail or succeed.
4) Less checks in the output
5) Production builds will additionally gain from single CI image
   pulled in order to perform the preparation of the packages
   and single package preparation step - it will save 4-5 minutes
   per image.

The disadvantage is a less clear output of such parallel build where
outputs from multiple builds will be interleaved in one CI output.

(cherry picked from commit 893d935ff01212085351b7ad13b19fca1d53bb57)

slyapustin committed:
Patch sql_alchemy_conn if old Postgres schemes used (#24569)

(cherry picked from commit 0527a0b6ce506434a23bc2a6f5ddb11f492fc614)

bbovenzi committed:
Migrate jsx files that affect run/task selection to tsx (#24509)

* convert all useSelection files to ts

Update grid data ts, remove some anys

* yarn, lint and tests

* convert statusbox to ts

* remove some anys, update instance tooltip

* fix types

* remove any, add comment for global vars

* fix url selection and grid/task defaults

* remove React.FC declarations

* specify tsconfig file path

* remove ts-loader

(cherry picked from commit c3c1f7ea2851377ba913075844a8dde9bfe6376e)

potiuk committed:
Upgrade FAB to 4.1.1 (#24399)

* Upgrade FAB to 4.1.1

The Flask Application Builder have been updated recently to
support a number of newer dependencies. This PR is the
attempt to migrate FAB to newer version.

This includes:

* update setup.py and setup.cfg upper and lower bounds to
  account for proper version of dependencies that
  FAB < 4.0.0 was blocking from upgrade
* added typed Flask application retrieval with a custom
  application fields available for MyPy typing checks.
* fix typing to account for typing hints added in multiple
  upgraded libraries optional values and content of request
  returned as Mapping
* switch to PyJWT 2.* by using non-deprecated "required" claim as
  list rather than separate fields
* add possibiliyt to install providers without constraints
  so that we could avoid errors on conflicting constraints when
  upgrade-to-newer-dependencies is used
* add pre-commit to check that 2.4+ only get_airflow_app is not
  used in providers
* avoid Bad Request in case the request sent to Flask 2.0 is not
  JSon content type
* switch imports of internal classes to direct packages
  where classes are available rather than from "airflow.models" to
  satisfy MyPY
* synchronize changes of FAB Security Manager 4.1.1 with our copy
  of the Security Manager.
* add error handling for a few "None" cases detected by MyPY
* corrected test cases that were broken by immutability of
  Flask 2 objects and better escaping done by Flask 2
* updated test cases to account for redirection to "path" rather
  than full URL by Flask2

Fixes: #22397

* fixup! Upgrade FAB to 4.1.1

(cherry picked from commit e2f19505bf3622935480e80bee55bf5b6d80097b)

potiuk committed:
Fix recording console for new rich-click 1.5 (#24611)

(cherry picked from commit 8231add0a0d12a1a5f2de13de7e4eb5ac785f839)

potiuk committed:
Add note about image regeneration in June 2022 (#24524)

The note is about image changes after refreshing are now better
organized (around date of the change) - this should be more useful
by the users who will look why their images have been refreshed.

Related to: #24516

(cherry picked from commit 1a62829136660b1c9b8e8a8676f4116f082cf7a2)

potiuk committed:
Fix errors revealed on autoupgrade of breeze (#24612)

Recent changes to Breeze cause it to fail in certain situations,
especially at self-upgrade (which was generated by today's
upgrade with rich-click).

There were two problems:

* docker volume inspect missed 'volume' and it caused sometimes
  failures in CI

* inputimeout dependency was missing after recent update to
  pre-commit venvs

(cherry picked from commit 0905e386f17e34d96f6ee575404c62b13242c75d)

kaxil committed:
Fix usage of `click.get_terminal_size()` (#24616)

We were ignoring mypy error instead of fixing it.

click had removed `get_terminal_size` and recommend using `shutil.get_terminal_size`

(cherry picked from commit 5f67cc0747ea661b703e4c44c77e7cd005cb9588)

potiuk committed:
Upgrade FAB to 4.1.2 (#24619)

(cherry picked from commit b8a4ac51f68a7d1c1f6ac5434cdf027e812553c7)

ashb committed:
We now need at least Flask-WTF 0.15 (#24621)

We upgraded flask and werkzeug in #24399, and updated the constraints,
but not everyone uses them (such as me in my local virtual environment
when developing) so the min version in setup.cfg has to match as well

(cherry picked from commit 030169dc3e5c17d8cade9c9a431afeee954abc3e)

itayB committed:
docs: sqlalchemy link update (#24627)

following to this change: https://github.com/apache/airflow/pull/22114

(cherry picked from commit 2113d72466fc89215c5874eab4124fce06a6d751)

potiuk committed:
Limit azure-servicebus to not be used on ARM (#24635)

Azure service bus uses uamqp which does not build for ARM architecture
and we need to disable it as a dependency for ARM.

(cherry picked from commit 477f907b347511a1b14053a418ba258db6c3ed99)

potiuk committed:
Allows to specify different Python base image when building images (#24634)

Today Python released new images that seems to break some of
our dependencies (at least on M1/ARM). This PR adds a workaround
possibility to add --python-image option to override the
default, latest image with any other version released previously
until we diagnose and fix the real issue.

(cherry picked from commit d6e6e7d2979e7a3c71d9b9a036630def0f9432f5)

potiuk committed:
Convert selective checks to Breeze Python (#24610)

Instead of bash-based, complex logic script to perform PR selective
checks we now integrated the whole logic into Breeze Python code.

It is now much simplified, when it comes to algorithm. We've
implemented simple rule-based decision tree. The rules describing
the decision tree are now are now much easier
to reason about and they correspond one-to-one with the rules
that are implemented in the code in rather straightforward way.

The code is much simpler and diagnostics of the selective checks
has also been vastly improved:

* The rule engine displays status of applying each rule and
  explains (with yellow warning message what decision was made
  and why. Informative messages are printed showing the resulting
  output

* List of files impacting the decision are also displayed

* The names of "ci file group" and "test type" were aligned

* Unit tests covering wide range of cases are added. Each test
  describes what is the case they demonstrate

* `breeze selective-checks` command that is used in CI can also
  be used locally by just providing commit-ish reference of the
  commit to check. This way you can very easily debug problems and
  fix them

Fixes: #19971
(cherry picked from commit d7bd72f494e7debec11672eeddf2e6ba5ef75fac)

potiuk committed:
Switch to new selective-checks in label-when-reviewed workflow (#24651)

When #24610 was implemented I missed the label-when-reviewed workflow

(cherry picked from commit 2703874eb27c85c781f8c20c8a62b3a4e8d8583b)

potiuk committed:
Cleanup references to selective checks (#24649)

Selective checks docs have been moved to breeze as part of #24610
but some of the references were still left.

This PR cleans it up.

(cherry picked from commit aa8cd30c46dde496423c72aa4a2c72b44d554745)

potiuk committed:
Remove misleading message from CI (#24650)

The CI jobs do not usually have mypy_cache volume created and
almost all jobs print misleading "mypy_cache_volume" error at
the beginning.

This is a noise - we are not interested in this stderr printed
message - we are only interested in returncode from checking
if it exists.

(cherry picked from commit ac104025f45b54efff26407e460a5e4279c6495b)

potiuk committed:
Remove selective checks from the "release workflow" (#24655)

Missed that one too :(

(cherry picked from commit 47f54b609983c6cb08553bb85245f2288deaf2dc)

potiuk committed:
Use target commit SHA for build image workflow (#24659)

The build-image workflow should use TARGET_COMMIT_SHA as the
selective check COMMIT_REF otherwise it might not build
image when needed

(cherry picked from commit 2ffaebee255c73e7a137b55ac25b95c16e159d75)

potiuk committed:
Handle "workflow_run" event properly in selective-check (#24656)

Unfortunately testing workflow_run is a bit difficult because
the changes are only effective after merging them.

Fixing (hopefully) yet another mistake in the workflow run where
commit hash was passed as event name (?)

We are going to handle "workflow_run" as valid event type so this
should now work without passing any event.

(cherry picked from commit e83e7c85c1bc3e634ad8a97224090028c0fd81b9)

potiuk committed:
Fix behaviour of build/pull after recent Breeze changes (#24657)

The behaviour of Breeze after some recent changes related to
pulling and building images in parallel have been slightly broken.
Nothing serious but slightly annoying behaviour:

* when starting breeze shell, the image was attempted to be
  build even if it was not needed (but cache efficiency made it fast
  enough to not be too annoying (unless we updated to newer
  python base image

* breeze pull command for "latest" branch makes no sense any more -
  we stopped pushing "latest" image to ghcr.io, we only push
  cache and "tagged" images. We are now turning --image-tag as
  required in "pull_image" and when someone specifies latest,
  error and helpful message is printed

* --force-build flag in "shell-related-commands" was not
  properly propagated to build-image so it did not actually
  force image building.

All those problems are fixed now.

(cherry picked from commit b678dc25f6604ddf603e42500e3e5583b6663df6)

potiuk committed:
Remove "Label when approved" workflow (#24704)

The labelling workflow has proven to be far less useful than we
thought and some of the recent changes in selective checks made
it largely obsolete. The committers can still add "full tests needed"
label when they think it is needed and there is no need to label
the PRs automatically for that (or any other reason).

For quite a while this workflow is basically a useless noise.

(cherry picked from commit 359700a450b9bf7afde5a2a6442dadadba36cd90)

bbovenzi committed:
don't try to render child rows for closed groups (#24637)

(cherry picked from commit 8f638bbeb9d3591a4e3c3000a542ae8edb58fb26)

bbovenzi committed:
Fix Grid vertical scrolling (#24684)

* fix vertical scrolling

* fix flex grow for panel open/close

* add type checking

* add duration axis component

* remove details/grid width changes

this should be done in a separate PR

(cherry picked from commit 1429091d7cf8e16e16efc7a0a3cc00f6ae5716a1)

potiuk committed:
Add ARM image building for regular PRs (#24664)

The image building for ARM is currently only done in the main build
only to refresh cache, however there are sometimes cases when
new dependency (for example #24635) broke ARM image build and it
was only discovered after merge.

This PR adds extra ARM-based build that should be run after
the AMD64 build. It should not influence the depending steps,
it should just signal failure of the PR if the ARM image cannot
be build.

(cherry picked from commit 2fbd750d35d12271ed4c5b9eb7564eb12e91bb03)

bbovenzi committed:
Debounce status highlighting in Grid view (#24710)

* Add delay / debounce to not always highlight tasks

* fix linting

* single delay variable at 200ms

(cherry picked from commit c7feb31786c7744da91d319f499d9f6015d82454)

potiuk committed:
Fix unnecessary check for ARM images (#24718)

The ARM image build introduced in #24664 had problem with build
image that was additionally checking for arm images which were
moved out to a spearate step

(cherry picked from commit 5321577aeef49a33d10efa42222df5524cc45c26)

potiuk committed:
Script to filter candidates for PR of the month based on heuristics (#24654)

This scripts proposes top candidates for PR of the month
based on simple heuristics as discussed in the document

https://docs.google.com/document/d/1qO5FztgzJLccfvbagX8DLh1EwhFVD2nUqbw96fRJmQQ/edit?disco=AAAAZ-Ct0Bs&usp_dm=true
(cherry picked from commit 0e1a6b98079814747205e3320e43e11f8e2ef3d4)

potiuk committed:
Merge-friendly output of command hashes for breeze (#24711)

We had just one hash generated from all commands in breeze and that
basically meant that when there were two PRs on two different
commands in Breeze, they resulted with merge conflict which
should be solved with `breeze regenerate-command-images`.

This change turns the hash output into a multi-command one - i.e.
each command has its own hash, which will make it much more
merge-friendly - i.e. if two PRs will work on two different commands
the rebase should result with merge rather than conflict.

(cherry picked from commit 94f4f81efb8c424bee8336bf6b8720821e48898a)

tanelk committed:
Do not fail requeued TIs (#23846)

(cherry picked from commit 66ffe39b0b3ae233aeb80e77eea1b2b867cc8c45)

tm-drtina committed:
Fix migration 0080_2_0_2 - Replace null values before setting column not null (#24585)

(cherry picked from commit 9f58e823329d525c0e2b3950ada7e0e047ee7cfd)

bbovenzi committed:
Address all yarn test warnings (#24722)

- remove outdated react-hooks library
- add missing mock

(cherry picked from commit c15cf6e10c7bdcb7389dd51f4b11650b7c930b0a)

potiuk committed:
Fix selective checks to work for non-main-branch

bbovenzi committed:
Fix grid date ticks (#24738)

* fix date ticks

* fix linting

* fix LinkButton type error

(cherry picked from commit d725625799d695f78e37696c3b594f18ccf2d899)

ephraimbuddy committed:
Update version to 2.3.3

ephraimbuddy committed:
Add release note for 2.3.3

jedcunningham commented on RELEASE_NOTES.rst:
```diff
@@ -21,7 +21,121 @@
+
+- We've upgraded Flask Application Builder to a major version 4.*.
+
+  Flask Application Builder is one of the important components of Airflow Webserver, as
```
```suggestion
- We've upgraded Flask App Builder to a major version 4.*.

  Flask App Builder is one of the important components of Airflow Webserver, as
```

They use "App", not "Application".

jedcunningham commented on RELEASE_NOTES.rst:
```diff
@@ -21,7 +21,121 @@
+  that have a number of security issues fixed. A lot of tests were performed to bring the dependencies
+  in a backwards-compatible way, however the dependencies themselves implement breaking changes in their
+  internals so it might be that some of those changes might impact the users in case they are using the
+  libraries for their onw purposes.
```
```suggestion
  libraries for their own purposes.
```

jedcunningham commented on RELEASE_NOTES.rst:
```diff
@@ -21,7 +21,121 @@
+
+  * Click 7.X to 8.X `breaking changes: <https://click.palletsprojects.com/en/8.0.x/changes/#version-8-0-0>`_
+
+  * itsdangerous 1.X to 2.X `breaking changes <https://github.com/pallets/itsdangerous/blob/main/CHANGES.rst#version-200>`_ (#24399)
```
```suggestion
  * itsdangerous 1.X to 2.X `breaking changes <https://github.com/pallets/itsdangerous/blob/main/CHANGES.rst#version-200>`_
```

Let's move this to the first line.

jedcunningham commented on RELEASE_NOTES.rst:
```diff
@@ -21,7 +21,121 @@
+Significant Changes
+^^^^^^^^^^^^^^^^^^^
+
+- We've upgraded Flask Application Builder to a major version 4.*.
```
```suggestion
- We've upgraded Flask Application Builder to a major version 4.*. (#24399)
```

jedcunningham commented on RELEASE_NOTES.rst:
```diff
@@ -21,7 +21,121 @@
+^^^^^^^^^
+
+- Allow for ``LOGGING_LEVEL=DEBUG`` (#23360)
+- Update Dockerfile (#24397)
```
```suggestion
```

jedcunningham commented on RELEASE_NOTES.rst:
```diff
@@ -21,7 +21,121 @@
+- docs: sqlalchemy link update (#24627)
+- Update ui.rst (#24514)
+- Update production-deployment.rst (#24121)
+- Fix inverted section levels in best-practices.rst (#23968)
```
```suggestion
```

I think we can skip all of these.

jedcunningham commented on RELEASE_NOTES.rst:
```diff
@@ -21,7 +21,121 @@
+- Add imports to deferring code samples (#24544)
+- Add note about image regeneration in June 2022 (#24524)
+- Small cleanup of ``get_current_context()`` chapter (#24482)
+- Docs: Fix default 2.2.5 log_id_template (#24455)
```
```suggestion
- Fix default 2.2.5 log_id_template (#24455)
```

jedcunningham commented on RELEASE_NOTES.rst:
```diff
@@ -21,7 +21,121 @@
+^^^^^^^^^^^^^
+
+- Remove internet explorer support (#24495)
+- Fix flaky order of returned dag runs (#24405)
```
I think this one might be categorized as a bug?

jedcunningham reviewed (commented):
Also missing the delete for the newsfragment.

collinmcnulty committed:
Document built in Timetables (#23099)

(cherry picked from commit 056b8cf3969137534f6f78b02166b71dd1b02ac6)

bbovenzi commented on RELEASE_NOTES.rst:
```diff
@@ -21,7 +21,121 @@
+- Debounce status highlighting in Grid view (#24710)
+- Fix Grid vertical scrolling (#24684)
+- don't try to render child rows for closed groups (#24637)
+- Restore tooltip animation (#24503)
```
```suggestion
- Restore tooltip animation (#24503)
```

This was fixing another commit here. Not a bug fix for release.

bbovenzi commented on RELEASE_NOTES.rst:
```diff
@@ -21,7 +21,121 @@
+- Fix toast messages (#24505)
+- Add indexes for CASCADE deletes for task_instance (#24488)
+- Return empty dict if Pod JSON encoding fails (#24478)
+- Remove ``framer-motion`` from custom tooltip (#24449)
```
```suggestion
- Remove ``framer-motion`` from custom tooltip (#24449)
```
We created the custom tooltip in 2.3.3 so we probably don't need to highlight this.

bbovenzi commented on RELEASE_NOTES.rst:
```diff
@@ -21,7 +21,121 @@
+Airflow 2.3.3 (2022-07-04)
+--------------------------
+
+Significant Changes
```
Should upgrading webpack or adding typescript to the UI be mentioned in the release notes?

ephraimbuddy replied:
I think the mention at `Misc/Internal` is enough unless there are extra actions users must take because of the change

fritz-astronomer committed:
Update templates doc to mention `extras` and format Airflow Vars / Conns (#24735)

(cherry picked from commit 54d5972cb87b7b0605d34a655cfe87b661fd75f1)

itayB committed:
docs: sqlalchemy link update (#24756)

(cherry picked from commit ca361ee1888cc8a0ffad3c916bb42b3c215c7919)

potiuk committed:
Fix selective checks to work for non-main-branch (#24769)

The recent changes in selective checks introduced a few problems
in non-main branches:

* constraints branch was not used by CI build in non-main branch
  this was not a proble, (until it started to fail) because
  other branches always work in "upgradeDepencies" case

* default branch calculation did not account for the selective-check
  comment that was run before default branch was retrieved. Therefore
  we did not run build as from 2.3 branch

* Some precomits should be skiped in non-main branch

* Integration was also excluded from Provider Check

* Two more commit outputs added:
  * docs-filter -  depends whether you are with more packages
  * skip-pre-commit - allows to dynamically choose which pre-commits
    should be skipped

(cherry picked from commit c8d3850fce4be19b998db35371657227f55ba6a9)

potiuk committed:
Choose the right builder when pushing to ghcr.io registry for cache (#24780)

When pushing cache we cannot use default builder because it has
no capability to push cache. We need to use airflow_cache for it,
which has additionally the capacity of forwarding the builds
to a remote ARM instance in case cache is built for ARM images
too. Recently we added a new --builder flag to allow to choose
the builder, but our scripts and CI have not been modified to
account for that (previously builder was chosen automatically but
that has proven to be limiting in some manual operations and it
also allows to choose different names for builders in case
someone wants to build their own pipeline of builds.

(cherry picked from commit a1679be85aa49c0d6a7ba2c31acb519a5bcdf594)

BasPH committed:
Align Black and blacken-docs configs (#24785)

(cherry picked from commit 46ac083f7e92cf8af046c8b5741e7c26ebffc05e)

jedcunningham commented on RELEASE_NOTES.rst:
```diff
@@ -21,7 +21,126 @@
+Significant Changes
+^^^^^^^^^^^^^^^^^^^
+
+- We've upgraded Flask App Builder to a major version 4.*. (#24399)
```
```suggestion
We've upgraded Flask App Builder to a major version 4.* (#24399)
""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
```

Sorry, just noticed this wasn't converted to a heading yet.

jedcunningham commented on RELEASE_NOTES.rst:
```diff
@@ -21,7 +21,126 @@
+
+    * ``Click`` 7.X to 8.X `breaking changes: <https://click.palletsprojects.com/en/8.0.x/changes/#version-8-0-0>`__
+
+    * ``itsdangerous`` 1.X to 2.X `breaking changes <https://github.com/pallets/itsdangerous/blob/main/CHANGES.rst#version-200>`__ (#24399)
```
```suggestion
    * ``itsdangerous`` 1.X to 2.X `breaking changes <https://github.com/pallets/itsdangerous/blob/main/CHANGES.rst#version-200>`__
```

jedcunningham commented on RELEASE_NOTES.rst:
```diff
@@ -21,7 +21,126 @@
+    * ``Click`` 7.X to 8.X `breaking changes: <https://click.palletsprojects.com/en/8.0.x/changes/#version-8-0-0>`__
+
+    * ``itsdangerous`` 1.X to 2.X `breaking changes <https://github.com/pallets/itsdangerous/blob/main/CHANGES.rst#version-200>`__ (#24399)
+
```
```suggestion
```

Extra newline

jedcunningham commented on RELEASE_NOTES.rst:
```diff
@@ -21,7 +21,126 @@
+- Fix flaky order of returned dag runs (#24405)
+- Migrate ``jsx`` files that affect run/task selection to ``tsx`` (#24509)
+- Fix links to sources for examples (#24386)
+
```
```suggestion
```

jedcunningham commented on RELEASE_NOTES.rst:
```diff
@@ -21,7 +21,126 @@
+- Fix default 2.2.5 log_id_template (#24455)
+- Update description of installing providers separately from core (#24454)
+- Mention context variables and logging (#24304)
+
```
```suggestion
```

jedcunningham commented on RELEASE_NOTES.rst:
```diff
@@ -21,7 +21,126 @@
+- Refactor ``DagRun.verify_integrity`` (#24114)
+- Upgrade FAB to 4.1.1 (#24399)
+- We now need at least Flask-WTF 0.15 (#24621)
+
```
```suggestion
```

ephraimbuddy replied:
I think we should have this extra space since it's the last before we get to 2.3.2

jedcunningham commented on RELEASE_NOTES.rst:
```diff
@@ -21,7 +21,126 @@
+- The JWT claims in the request to retrieve logs have been standardized: we use ``nbf`` and ``aud`` claims for
+  maturity and audience of the requests. Also "filename" payload field is used to keep log name. (#24519)
+- Address all ``yarn`` test warnings (#24722)
+- Add tests for the ``grid_data`` endpoint (#24375)
```
This should probably be a bugfix and reworded, something like "Set proper Content-Type and chartset on ``grid_data`` endpoint".

jedcunningham commented on RELEASE_NOTES.rst:
```diff
@@ -21,7 +21,126 @@
+  maturity and audience of the requests. Also "filename" payload field is used to keep log name. (#24519)
+- Address all ``yarn`` test warnings (#24722)
+- Add tests for the ``grid_data`` endpoint (#24375)
+- Remove ``framer-motion`` from custom tooltip (#24449)
```
Why isn't this a bugfix?

ephraimbuddy replied:
Good catch. I will combine it with `Improve grid rendering performance with a custom tooltip (#24417)`,

potiuk committed:
Unified "dash-name" convention for outputs in ci workflows.

There were errors with retieving constraints branch caused by
using different convention for output names (sometimes dash,
sometimes camelCase as suggested by most GitHub documents).

The "dash-name" looks much better and is far more readable so
we shoud unify all internal outputs to follow it.

During that rename some old, unused outputs were removed,
also it turned out that the new selective-check can
replace previous "dynamic outputs" written in Bash as well.

Additionally, the "defaults" are now retrieved via Python script, not
bash script which will make it much more readable - both build_images
and ci.yaml use it in the right place - before replacing
the scripts and dev with the version coming in from PR in case
of build_images.yaml.

(cherry picked from commit 017507be1e1dbf39abcc94a44fab8869037893ea)
